src-cpp/rdkafkacpp.h

```diff
@@ -1365,7 +1367,267 @@ class RD_EXPORT MessageTimestamp {
   int64_t timestamp;               /**< Milliseconds since epoch (UTC). */
 };
 
+/**
+ * @brief Headers object
+ *
+ * This object encapsulates the C implementation logic into a C++ object
+ * for use in RdKafka::Messages object.
+ * 
+ * @remark Requires Apache Kafka >= 0.11.0 brokers
+ */
+class RD_EXPORT Headers {
+ public:
+  virtual ~Headers() = 0;
+
+  /**
+   * @brief Header object
+   *
+   * This object represents a single Header with key value pair
+   * and an ErrorCode
+   *
+   * @remark dynamic allocation of this object is not supported.
+   */
+  class Header {
+   public:
+    /**
+     * @brief Header object to encapsulate a single Header
+     *
+     * @param key the string value for the header key
+     * 
+     * @param value the bytes of the header value
+     * 
+     * @param value_size the length in bytes of the header value
+     */
+    Header(const std::string &key,
+           const void *value,
+           size_t value_size):
+    key_(key), err_(RdKafka::ERR_NO_ERROR), value_size_(value_size) {
+        value_ = copy_value(value, value_size);
+    };
+
+    /**
+     * @brief Header object to encapsulate a single Header
+     *
+     * @param key the string value for the header key
+     * 
+     * @param value the bytes of the header value
+     * 
+     * @param value_size the length in bytes of the header value
+     * 
+     * @param err the error code if one returned
+     *
+     * @remark The error code is used for when the Header is constructed internally
+     *         by using something like RdKafka::Headers::get_last which constructs
+     *         a Header encapsulating the ErrorCode in the process
+     */
+    Header(const std::string &key,
+           const void *value,
+           size_t value_size,
+           const RdKafka::ErrorCode &err):
+    key_(key), err_(err), value_size_(value_size)  {
+        value_ = copy_value(value, value_size);
```

Comment:
It would be more efficient to just reference the underlying C header value data, if possible